Firestore/fuzzing/leveldb_fuzzer.cc

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cstddef>
+#include <cstdint>
+#include <string>
+
+#include "Firestore/core/src/firebase/firestore/local/leveldb_key.h"
+#include "Firestore/core/src/firebase/firestore/local/leveldb_util.h"
+
+using firebase::firestore::local::LevelDbDocumentMutationKey;
+using firebase::firestore::local::LevelDbDocumentTargetKey;
+using firebase::firestore::local::LevelDbMutationKey;
+using firebase::firestore::local::LevelDbMutationQueueKey;
+using firebase::firestore::local::LevelDbQueryTargetKey;
+using firebase::firestore::local::LevelDbRemoteDocumentKey;
+using firebase::firestore::local::LevelDbTargetDocumentKey;
+using firebase::firestore::local::LevelDbTargetGlobalKey;
+using firebase::firestore::local::LevelDbTargetKey;
+using firebase::firestore::model::ResourcePath;
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+  const char* str_ptr = reinterpret_cast<const char*>(data);
+  absl::string_view str{str_ptr, size};
+  leveldb::Slice slice = firebase::firestore::local::MakeSlice(str);
+
+  // Test DescribeKey(std::string) which calls MakeSlice(std::string).
+  try {
+    firebase::firestore::local::DescribeKey(str);
+  } catch (...) {
+    // Ignore caught errors and assertions.
+  }
+
+  // Test LevelDbMutationKey methods.
+  try {
+    LevelDbMutationKey::KeyPrefix(str);
+  } catch (...) {
+    // Ignore caught errors and assertions.
+  }
+
+  try {
+    LevelDbMutationKey::Key(str, static_cast<int>(size));
```

Comment:
This is a little misleading. The second parameter is the batch_id, and has nothing to do with the size.

Possible solutions:
1) Cast size to a model::BatchId rather than an int. This at least highlights that the second parameter isn't size related, and that you're just using the size as the input for lack of something better. But you could theoretically run into an issue where the size of the input somehow masked an issue related to the batchid? It might be better to keep batchid independent from size, leading to:
2) Use the first 32 bits of data as the batch id, and the remainder as str. This requires a bit more juggling of the input, and you might decide that it's not worth it in this case.

Either is fine by me.

Reply:
I casted the `size` to an `int` then used it to create a `BatchId` object. I think this will only cause problems if the casting from `size_t` to `int` fails